### PR TITLE
Generate Asp.Net extensions

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/2762
Fixes https://github.com/Azure/azure-sdk-for-net/issues/34074

# Description

Guidelines: https://azure.github.io/azure-sdk/dotnet_implementation.html#integration-with-aspnet-core

The generator will generate an extension for `IAzureClientFactoryBuilderWithCredential` when a constructor of the client takes a `TokenCredential` as parameter, generate an extension for `IAzureClientFactoryBuilder` instead when the constructor does not take `TokenCredential` as parameter.
The generator will always generate an extension method for `IAzureClientFactoryBuilderWithConfiguration<TConfiguration>`

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/LowLevel/Output/AspDotNetExtension.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+using AutoRest.CSharp.Output.Models;
+using AutoRest.CSharp.Output.Models.Shared;
+using AutoRest.CSharp.Output.Models.Types;
+using AutoRest.CSharp.Utilities;
+using Azure.Core;
+using Azure.Core.Extensions;
+
+namespace AutoRest.CSharp.LowLevel.Output
+{
+    internal class AspDotNetExtension : TypeProvider
```
Comment (pshao25) [inline src/AutoRest.CSharp/LowLevel/Output/AspDotNetExtension.cs:21]: It is not AspDotNet specific, I would call it `ClientExtesion` or `ServiceCollectionExtenstion`

File: test/CadlRanchProjects/authentication/api-key/Generated/AuthenticationApiKeyClientBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        /// <param name="credential"> A credential used to authenticate to an Azure Service. </param>
+        /// <param name="endpoint"> TestServer endpoint. </param>
+        public static IAzureClientBuilder<ApiKeyClient, ApiKeyClientOptions> AddApiKeyClient<TBuilder>(this TBuilder builder, AzureKeyCredential credential, Uri endpoint)
+        where TBuilder : IAzureClientFactoryBuilder
+        {
+            return builder.RegisterClientFactory<ApiKeyClient, ApiKeyClientOptions>((options) => new ApiKeyClient(credential, endpoint, options));
+        }
+
+        /// <param name="builder"></param>
+        /// <param name="configuration"></param>
```
Comment (pshao25) [inline test/CadlRanchProjects/authentication/api-key/Generated/AuthenticationApiKeyClientBuilderExtensions.cs:28]: No content of comments?

File: test/CadlRanchProjects/models/property-optional/Generated/ModelsPropertyOptionalClientBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        /// <param name="builder"></param>
+        /// <param name="configuration"></param>
+        public static IAzureClientBuilder<OptionalClient, OptionalClientOptions> AddOptionalClient<TBuilder, TConfiguration>(this TBuilder builder, TConfiguration configuration)
+        where TBuilder : IAzureClientFactoryBuilderWithConfiguration<TConfiguration>
+        {
+            return builder.RegisterClientFactory<OptionalClient, OptionalClientOptions>(configuration);
+        }
+        /// <param name="builder"></param>
+        /// <param name="configuration"></param>
+        public static IAzureClientBuilder<Models.Property.Optional.String, OptionalClientOptions> AddString<TBuilder, TConfiguration>(this TBuilder builder, TConfiguration configuration)
```
Comment (pshao25) [inline test/CadlRanchProjects/models/property-optional/Generated/ModelsPropertyOptionalClientBuilderExtensions.cs:34]: This looks weird, and I even wonder whether if it works. But I think we could create an issue about the subclient scenarios and figure it out later.

Comment (pshao25) [reply test/CadlRanchProjects/models/property-optional/Generated/ModelsPropertyOptionalClientBuilderExtensions.cs:34]: Or maybe we just not generate it for now.

File: src/AutoRest.CSharp/LowLevel/Output/AspDotNetExtension.cs (hunk 1 above)
Comment (m-nash) [reply src/AutoRest.CSharp/LowLevel/Output/AspDotNetExtension.cs]: It is aspdotnet specific

File: test/CadlRanchProjects/models/property-optional/Generated/ModelsPropertyOptionalClientBuilderExtensions.cs (hunk 3 above)
Comment (m-nash) [reply test/CadlRanchProjects/models/property-optional/Generated/ModelsPropertyOptionalClientBuilderExtensions.cs:37]: OperationGroups and subclients should be excluded here.

**Review by m-nash — APPROVED** [review]

**Pull request merged**